kube-controller-manager panic when network interfaces for Azure VMSS are null

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:

kube-controller-manager panic when network interfaces for Azure VMSS are null. This happens when Azure cloud provider tried to get IP address for VMSS VM:

```
I0823 22:21:23.235856       1 route_controller.go:176] Creating route for node aks-node-1 10.244.7.0/24 with hint b5759222-0000-45dd-9ecd-11110683608f, throttled 800ns
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x22335bd]
 
goroutine 257858 [running]:
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*scaleSet).getPrimaryInterfaceID(0xc000e2bcb0, 0x0, 0xc0053a97d0, 0xc001236d80, 0xc0048ebd00, 0x0, 0xc0012371e0, 0x0, 0xc0053a97e0, 0xc0053a96d0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_vmss.go:443 +0x2d
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*scaleSet).GetPrimaryInterface(0xc000e2bcb0, 0xc004617b90, 0x23, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_vmss.go:657 +0x52e
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*scaleSet).GetIPByNodeName(0xc000e2bcb0, 0xc004617b90, 0x23, 0xc005a66480, 0x30, 0x30, 0x509e9e0, 0xc000069470, 0xc000069400)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_vmss.go:382 +0x80
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).GetIPForMachineWithRetry.func1(0x509e9e0, 0xc003fab901, 0xc005a66480)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_backoff.go:139 +0x79
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x0, 0x0, 0x0, 0x1, 0x0, 0xc005a66480, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:284 +0x51
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).GetIPForMachineWithRetry(0xc0009f6f00, 0xc004617b90, 0x23, 0x5333ec0, 0xc005a66450, 0xc004617b90, 0x23, 0xc375220, 0x433200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_backoff.go:137 +0x13e
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).getIPForMachine(0xc0009f6f00, 0xc004617b90, 0x23, 0x443700, 0x0, 0x0, 0xc003fabad8, 0x443bed, 0xc005ba5ec0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_backoff.go:131 +0xe3
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).CreateRoute(0xc0009f6f00, 0x88d54e0, 0xc00005c050, 0x5516421, 0xa, 0xc004617bc0, 0x24, 0xc003bedb00, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_routes.go:286 +0x2b8
k8s.io/kubernetes/pkg/controller/route.(*RouteController).reconcile.func1.1(0x7fb7008f0b28, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/route/route_controller.go:177 +0x2d4
k8s.io/kubernetes/vendor/k8s.io/client-go/util/retry.RetryOnConflict.func1(0x4e76a20, 0xc003b86601, 0xc0046ffc20)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/retry/util.go:64 +0x33
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x5f5e100, 0x0, 0x3ff0000000000000, 0x5, 0x0, 0xc0046ffc20, 0x40fed8, 0x40)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:284 +0x51
k8s.io/kubernetes/vendor/k8s.io/client-go/util/retry.RetryOnConflict(0x5f5e100, 0x0, 0x3ff0000000000000, 0x5, 0x0, 0xc003bedb40, 0x1, 0xc00426dcb0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/retry/util.go:63 +0xa6
k8s.io/kubernetes/pkg/controller/route.(*RouteController).reconcile.func1(0xc002f94bd0, 0xc002219d40, 0xc0003d41c0, 0xc004617b90, 0x23, 0xc004617bc0, 0x24, 0xc003bedb00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/route/route_controller.go:171 +0x123
created by k8s.io/kubernetes/pkg/controller/route.(*RouteController).reconcile
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/route/route_controller.go:169 +0x38c
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x22335bd]
```

**What you expected to happen**:

kube-controller-manager should report an error instead of panic.

**How to reproduce it (as minimally and precisely as possible)**:

Scale cluster nodes up and down frequently.

**Anything else we need to know?**:

/kind bug
/area provider/azure

**Environment**:
- Kubernetes version (use `kubectl version`): 1.15+
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

--- a/staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go
+++ b/staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go
@@ -534,17 +534,21 @@ func (ss *scaleSet) GetPrivateIPsByNodeName(nodeName string) ([]string, error) {
 
 // This returns the full identifier of the primary NIC for the given VM.
 func (ss *scaleSet) getPrimaryInterfaceID(machine compute.VirtualMachineScaleSetVM) (string, error) {
+	if machine.NetworkProfile == nil || machine.NetworkProfile.NetworkInterfaces == nil {
+		return "", fmt.Errorf("failed to find the network interfaces for vm %s", to.String(machine.Name))
+	}
+
 	if len(*machine.NetworkProfile.NetworkInterfaces) == 1 {
 		return *(*machine.NetworkProfile.NetworkInterfaces)[0].ID, nil
 	}
 
 	for _, ref := range *machine.NetworkProfile.NetworkInterfaces {
-		if *ref.Primary {
+		if to.Bool(ref.Primary) {
 			return *ref.ID, nil
 		}
 	}
 
-	return "", fmt.Errorf("failed to find a primary nic for the vm. vmname=%q", *machine.Name)
+	return "", fmt.Errorf("failed to find a primary nic for the vm. vmname=%q", to.String(machine.Name))
 }
 
 // getVmssMachineID returns the full identifier of a vmss virtual machine.

--- a/staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss_test.go
+++ b/staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss_test.go
@@ -949,6 +949,11 @@ func TestGetPrimaryInterfaceID(t *testing.T) {
 			},
 			expectedErr: fmt.Errorf("failed to find a primary nic for the vm. vmname=\"vm\""),
 		},
+		{
+			description:       "GetPrimaryInterfaceID should report an error if there's no network interface on the VMSS VM",
+			existedInterfaces: []compute.NetworkInterfaceReference{},
+			expectedErr:       fmt.Errorf("failed to find the network interfaces for vm vm"),
+		},
 	}
 
 	for _, test := range testCases {
@@ -962,6 +967,9 @@ func TestGetPrimaryInterfaceID(t *testing.T) {
 					NetworkInterfaces: &test.existedInterfaces,
 				},
 			},
+		}
+		if len(test.existedInterfaces) == 0 {
+			vm.VirtualMachineScaleSetVMProperties.NetworkProfile = nil
 		}
 
 		id, err := ss.getPrimaryInterfaceID(vm)

